crates/matrix-sdk/src/encryption/recovery/mod.rs

```diff
@@ -0,0 +1,526 @@
+        }
+    }
+
+    /// Disable recovery completely.
+    ///
+    /// This method will do the following steps:
+    ///
+    /// 1. Disable the uploading of room keys to a currently active backup.
+    /// 2. Delete the currently active backup.
+    /// 3. Set the `m.secret_storage.default_key` global account data event to
+    ///    an empty JSON content.
+    /// 4. Set a global account data event so clients won't attempt to
+    ///    automatically re-enable a backup.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// recovery.disable().await?;
+    ///
+    /// assert_eq!(recovery.state(), RecoveryState::Disabled);
+    ///
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub async fn disable(&self) -> Result<()> {
+        self.client.encryption().backups().disable().await?;
+        // Why oh why, can't we delete account data events?
+        self.client.account().set_account_data(SecretStorageDisabledContent {}).await?;
+        self.client.account().set_account_data(BackupDisabledContent { disabled: true }).await?;
+        self.update_recovery_state().await?;
+        // TODO: Do we want to "delete" the known secrets as well?
+
+        Ok(())
+    }
+
+    /// Reset the recovery key.
+    ///
+    /// This will rotate the secret storage key and re-upload all the secrets to
+    /// the [`SecretStore`].
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// let new_recovery_key =
+    ///     recovery.reset_key().with_passphrase("my passphrase").await;
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub fn reset_key(&self) -> Reset<'_> {
+        // TODO: Should this only be possible if we're in the RecoveryState::Enabled
+        // state? Otherwise we'll create a new secret store but won't be able to
+        // upload all the secrets.
+        Reset::new(self)
+    }
+
+    /// Reset the recovery key but first import all the secrets from secret
+    /// storage.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
```

Comment:
… and here.

Reply:
Same 77444856aab8b321070293c3d661a4efa2f33082